docs/specs/verification.md

```diff
@@ -0,0 +1,66 @@
+# User Verification
+
+## Definitions
+
+A user is verified if the user fulfill the condition specified by criteria.
+
+An identity is verifiable if:
+- it is an SSO identity; or
+- it is a login ID identity with a configured verifiable login ID key.
+
+An identity is verified if:
+- it is an SSO identity; or
+- it is a login ID identity, and the user has a matching OOB-OTP authenticator.
+
+## Criteria
+
+Developer can configure the criteria used to determine verification status
+of a user. There are two possible criteria:
+
+- `any`: User has a verifiable identity and one of user's verifiable identities is verified.
+- `all`: User has a verifiable identity and all of user's verifiable identities are verified.
+
+By default, criteria `any` is used.
+
+```yaml
+# Use 'all' criteria to determine user verification status
+verification:
+    criteria: all
+```
+
+## Requirement
+
+Developer can configure verification requirement for specific login ID keys.
+The specified login ID keys must have type `email` or `phone`.
+
+When a login ID identity with specified login ID key is created (e.g. during
+sign up), the user is required to verify the login ID using a one-time-
+password sent to the login ID before proceeding. A matching OOB-OTP
+authenticator would be created in the verification process.
+
+By default, user must verify login ID key 'email' and 'phone'.
+
+```yaml
+# Require verification for login ID key 'email' and 'phone'
+verification:
+    login_id_keys: [email, phone]
+```
+
+## Code & Message
+
+The verification code & message can be configured using OOB-OTP authenticator
```

Comment:
Does it meant, if a user have two ways to login:

1. Email + Password
2. Email + OTP

(p.s. basically this is the case of Slack by default, after user input email address, they can choose to login manually or send a "Magic link")

The verification email after sign up, and the OTP login email content, will be the same? If yes we might have a problem.

Reply:
> The verification email after sign up, and the OTP login email content, will be the same?

Let me try to clarify your use case:
- User signup with email + password, no email verification is needed
- User can verify email in settings page
- User can login with email + password, or email + OTP if email is verified

If that's the case, I suppose we need to allow different content of the OTP message depending on the context.

If so, we can provide template entries for OTP messages (`otp_message_email.txt`/`otp_message_email.html`/`otp_message_sms.txt`), and put the originating operation type (`operation` = `verification` or `authentication`) in the template context. Does that sounds good to you?

Reply:
If we take Slack as the sample, the use case is more like:

1. User signup with email + password, and verify email address
2. User can login with email + password, or email + OTP

Your solution sounds ok, but I'm not sure what is the user experience will be if we do operation type vs more template entries... can you / @louischan-oursky advice?

Reply:
Discussed with @louischan-oursky , we agree that having just one template type for OTP message is better in term of developer experience, since they just need to customize single set file, versus copy & paste a dozen of template files and change just some words.

We can add some fields to indicate the context of the message. For example:
- source: signup/login/settings
- operation: primary_auth/secondary_auth/verify

Let me update the specification to reflect this.